### PR TITLE
encode leader info from heartbeat, update show hosts/parts

### PR DESCRIPTION
**Pull request by critical27**
Follow-up of #1236, just encode all leader info into a new table `leaders`, and when we show hosts/parts, we can use this table instead of contacting with storage service.

After this merged, we can use it to preheat storage client.

Known issues: because we judge whether a host is active by heartbeat, if we kill a storage node, it would seems we have `more` leader than we should have. See examples below: `heartbeat interval` is 3s, `expired_threshold_sec` is 30s, total leader count should be 100.

1. Normal state
image

2. Now we kill node 27006, we will get extra 35 leaders, which is the leader count of 27006 before, however the first nodes leader count sum is 100.
image

3. When pass `expired_threshold_sec`, the 27006 is regarded as `offline`, everything looks good.
image

### CODE REVIEW

**Comment by dangleptr** [comment]
Well done.  
For the problem you mentioned,  maybe we could filter the host which lastHeartBeatTime expired with heartbeat_interval as the threshold

WDYT?

**Comment by critical27** [comment]
> Well done.
> For the problem you mentioned, maybe we could filter the host which lastHeartBeatTime expired with heartbeat_interval as the threshold when showing hosts.
> 
> WDYT?

It that a little bit vulnerable to network? If we lost one heartbeat, leader count would be zero. Let me have a try.

File: src/meta/processors/partsMan/ListHostsProcessor.cpp [hunk 1]
```diff
@@ -56,43 +55,52 @@ StatusOr<std::vector<cpp2::HostItem>> ListHostsProcessor::allHostsWithStatus(
...
         iter->next();
     }
 
     std::unordered_map<HostAddr,
                        std::unordered_map<std::string, std::vector<PartitionID>>> allParts;
-    for (const auto& spaceId : spaces) {
+    for (const auto& spaceId : spaceIds_) {
         // get space name by space id
-        auto spaceName = spaceIdNameMap[spaceId];
+        auto spaceName = spaceIdNameMap_[spaceId];
```
Comment (wadeliuyi) [inline src/meta/processors/partsMan/ListHostsProcessor.cpp]: const auto &spaceName?

File: src/meta/processors/partsMan/ListPartsProcessor.cpp [hunk 2]
```diff
@@ -91,34 +96,28 @@ ListPartsProcessor::getAllParts() {
...
-                } else if (showAllParts_ || std::find(partIds_.begin(), partIds_.end(), partId) ==
-                                            partIds_.end()) {
-                    LOG(ERROR) << "Maybe not get the leader of partition " << partId;
+    auto activeHosts = ActiveHostsMan::getActiveHosts(kvstore_);
+    while (iter->valid()) {
+        auto host = MetaServiceUtils::parseLeaderKey(iter->key());
+        if (std::find(activeHosts.begin(), activeHosts.end(),
+                      HostAddr(host.ip, host.port)) != activeHosts.end()) {
+            LeaderParts leaderParts = MetaServiceUtils::parseLeaderVal(iter->val());
+            auto partIds = leaderParts[spaceId_];
```
Comment (wadeliuyi) [inline src/meta/processors/partsMan/ListPartsProcessor.cpp]: const auto &partIds?

File: src/meta/processors/partsMan/ListPartsProcessor.cpp [hunk 3]
```diff
@@ -91,34 +96,28 @@ ListPartsProcessor::getAllParts() {
...
-                    it->set_leader(this->toThriftHost(leader));
-                } else if (showAllParts_ || std::find(partIds_.begin(), partIds_.end(), partId) ==
-                                            partIds_.end()) {
-                    LOG(ERROR) << "Maybe not get the leader of partition " << partId;
+    auto activeHosts = ActiveHostsMan::getActiveHosts(kvstore_);
+    while (iter->valid()) {
+        auto host = MetaServiceUtils::parseLeaderKey(iter->key());
+        if (std::find(activeHosts.begin(), activeHosts.end(),
+                      HostAddr(host.ip, host.port)) != activeHosts.end()) {
+            LeaderParts leaderParts = MetaServiceUtils::parseLeaderVal(iter->val());
```
Comment (wadeliuyi) [inline src/meta/processors/partsMan/ListPartsProcessor.cpp]: each loop will create a template map, and will destruct after the loop, maybe here can define a function named parseLeaderValBySpaceID(iter->val(), spaceId_) to just return the leaders of current space.

File: src/meta/processors/admin/HBProcessor.cpp [hunk 4]
```diff
@@ -37,25 +37,32 @@ void HBProcessor::process(const cpp2::HBReq& req) {
...
-        ret = ActiveHostsMan::updateHostInfo(kvstore_, host, info);
         if (req.__isset.leader_partIds) {
-            const auto& leaderIds = req.get_leader_partIds();
-            UNUSED(leaderIds);
-            // TODO(zhangguoqing) put the leader parts information into kvstore
+            ret = ActiveHostsMan::updateHostInfo(kvstore_, host, info,
+                                                 req.get_leader_partIds());
+            updateLeader = true;
+            lastUpdateTime = time::WallClock::fastNowInMilliSec();
+            LastUpdateTimeMan::update(kvstore_, lastUpdateTime);
```
Comment (dangleptr) [inline src/meta/processors/admin/HBProcessor.cpp]: It seems that we don't need to update the lastUpdateTime in HeartBeat processor.

File: src/meta/processors/partsMan/ListHostsProcessor.cpp (hunk 1 above)
Comment (critical27) [reply src/meta/processors/partsMan/ListHostsProcessor.cpp]: I remeber that operator[] return T&?

File: src/meta/processors/partsMan/ListPartsProcessor.cpp (hunk 3 above)
Comment (critical27) [reply src/meta/processors/partsMan/ListPartsProcessor.cpp]: Good suggestion

File: src/meta/processors/admin/HBProcessor.cpp (hunk 4 above)
Comment (critical27) [reply src/meta/processors/admin/HBProcessor.cpp]: Indeed, fixed

**Comment by dutor** [comment]
Jenkins, go

File: src/meta/processors/admin/HBProcessor.cpp [hunk 5]
```diff
@@ -40,20 +40,20 @@ void HBProcessor::process(const cpp2::HBReq& req) {
...
         HostInfo info(time::WallClock::fastNowInMilliSec());
-        ret = ActiveHostsMan::updateHostInfo(kvstore_, host, info);
         if (req.__isset.leader_partIds) {
-            const auto& leaderIds = req.get_leader_partIds();
-            UNUSED(leaderIds);
-            // TODO(zhangguoqing) put the leader parts information into kvstore
+            ret = ActiveHostsMan::updateHostInfo(kvstore_, host, info,
+                                                 req.get_leader_partIds());
+        } else {
+            ret = ActiveHostsMan::updateHostInfo(kvstore_, host, info);
```
Comment (dangleptr) [inline src/meta/processors/admin/HBProcessor.cpp]:
```
ret = ActiveHostsMan::updateHostInfo(
    kvstore_, 
    host, 
    info,  
    req.__isset.leader_partIds ? req.get_leader_partIds() : nullptr);
```

**Pull request merged**